### PR TITLE
fix: NM Settings leak when active connection does not exist

### PR DESCRIPTION
**Pull request by GregoryIvo**
This PR modifies the [GetAppliedConnection()](https://networkmanager.dev/docs/api/latest/gdbus-org.freedesktop.NetworkManager.Device.html#gdbus-method-org-freedesktop-NetworkManager-Device.GetAppliedConnection) method.

Old behaviour: would find an associated connection and update an existing one. If it doesn’t it creates a new one.

New Behaviour: check for associated connections, then check for non-active associated connections. If more than one associated connection is found it will take the first and delete the rest. If none is found, create a new one.

> **Note**: We are using the Conventional Commits convention for our pull request titles. Please take a look at the [PR title format document](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#submitting-the-changes) for the supported [types](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#type) and [scopes](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#scope).

Brief description of the PR. [e.g. Added `null` check on `object` to avoid `NullPointerException`]

**Related Issue:** This PR fixes/closes {issue number}

**Description of the solution adopted:** A more detailed description of the changes made to solve/close one or more issues. If the PR is simple and easy to understand this section can be skipped.

**Screenshots:** If applicable, add screenshots to help explain your solution

**Any side note on the changes made:** Description of any other change that has been made, which is not directly linked to the issue resolution [e.g. Code clean up/Sonar issue resolution]

### CODE REVIEW

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java [hunk 1]
```diff
@@ -443,8 +443,50 @@ private Optional<Connection> getAppliedConnection(Device dev) throws DBusExcepti
...
+            for (DBusPath path : connectionPath) {
+
+                Connection workingConnection = this.dbusConnection.getRemoteObject(NM_BUS_NAME, path.getPath(), Connection.class);
+                
+                Map<String, Map<String, Variant<?>>> workingSettings = workingConnection.GetSettings();
+                
+                String workingConnectionName = (String) workingSettings.get("connection").get("id").getValue();
+                String workingConnectionId = (String) workingSettings.get("connection").get("uuid").getValue();
+                
+                if (workingConnectionName.contains(targetConnectionName)) {
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java]: We're looking for an exact match, not partial

```suggestion
                if (workingConnectionName.equals(targetConnectionName)) {
```

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java [hunk 2]
```diff
@@ -443,8 +443,50 @@ private Optional<Connection> getAppliedConnection(Device dev) throws DBusExcepti
...
+                    if (isFirst) {
+                        logger.info("Using avaliable connection uuid: {}", workingConnectionId);
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java]: Lower the log level a little:

```suggestion
                        logger.debug("Using avaliable connection uuid: {}", workingConnectionId);
```

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java [hunk 3]
```diff
@@ -443,8 +443,50 @@ private Optional<Connection> getAppliedConnection(Device dev) throws DBusExcepti
...
+                        connectionToReturn = Optional
+                                .of(workingConnection);
+                        isFirst = false;
+                    } else {
+                        logger.info("Deleting extra connection with uuid: {}", workingConnectionId);
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java]:
```suggestion
                        logger.debug("Deleting extra connection with uuid: {}", workingConnectionId);
```

**Review by mattdibi — CHANGES_REQUESTED** [review]
Couple of notes:
- The variable names don't help reading the code.
  - `workingConnection`->`availableConnection` is better.
  - `workingConnectionName`->`availableConnectionId`
  - `workingConnectionId`->`availableConnectionUuid`
  - `targetConnectionName` -> `expectedConnectionName` maybe?
-  **Tests**! Add the proper unit tests for this new behaviour. I expect to find test cases that only target this (i.e. don't change the existing ones, create new test cases) and cover all the scenarios (no available connection, multiple available connection, multiple matching connection etc etc).

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java (hunk 1 above)
Comment (GregoryIvo) [reply kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java]: Will do! 

My thought process was that we discussed potentially adding an extra identifier to the name.

**Comment by GregoryIvo** [comment]
@mattdibi  kindly re-review this for me tomorrow! Thank you

**Review by mattdibi — CHANGES_REQUESTED** [review]
@GregoryIvo build fails

image

**Comment by GregoryIvo** [comment]
> @GregoryIvo build fails
> 
> 
> 
> image

Uggh it must be because of my manual merge attempt via GitHub gui 😪

**Comment by MMaiero** [comment]
@GregoryIvo Can we try to solve this? It would be great to have it merged and close for the review

**Comment by GregoryIvo** [comment]
I currently on it!

**Comment by GregoryIvo** [comment]
This PR has been rebased in #4477

**Comment by mattdibi** [comment]
Closing as superseded by #4477

**Pull request closed without merging**